### PR TITLE
There is mine.

### PR DESCRIPTION
**Pull request by yiwan**
Maybe I take too long.
But it is a good try of backend development.

### CODE REVIEW

**Comment by makara** [comment]
Thanks. We'll review and get back to you soon.

File: lightning_talk/views/post.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    @detail_route(url_path='upvote', methods=['PUT', 'DELETE'])
+    def upvote(self, request, pk=None):
+        post = self.get_object()
+        exists = post.upvotes.filter(user=self.request.user).exists()
+
+        if self.request.method == 'PUT':
+            if exists:
+                return JsonResponse(
+                    data={'detail': 'already upvoted'}, status='500')
```
Comment (JuhaS) [inline lightning_talk/views/post.py]: I think 500 is not the correct status code for this, can you recommend a better one for this kind of case? Same for line 47 of the same file.

File: lightning_talk/models/post.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+from django.conf import settings
+from .upvote import Upvote
+
+
+class Post(models.Model):
+    title = models.CharField(max_length=40)
+    description = models.CharField(max_length=200)
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.SET_NULL, null=True)
+    upvotes = models.ManyToManyField(Upvote, blank=True)
```
Comment (JuhaS) [inline lightning_talk/models/post.py]: I think there is no need to make Post <> Upvote relation ManyToMany. One upvote will only belong to single Post. Can you update to use ForeignKey relation instead? Or was there some other reason to use ManyToMany?

**Comment by JuhaS** [comment]
Hi, @yiwan I tested it out and worked without problems. I added few comments to the PR for some backend, can you check and let me know what you think?

File: lightning_talk/models/post.py (hunk 2 above)
Comment (yiwan) [inline lightning_talk/models/post.py]: What I want was OneToMany, Post contains different users' Upvote or none. It seems like Django has no built-in OneToMany, so I use ManyToMany, works fine.
Maybe set Post as Upvote's ForeignKey will be better, but I need write more code.

**Comment by yiwan** [comment]
@JuhaS Just let you know I have made some change to code, and explained why I use Django's `ManyToMany ` field.

File: lightning_talk/models/post.py (hunk 2 above)
Comment (JuhaS) [inline lightning_talk/models/post.py]: The way to represent one-to-many relations in Django is to use ForeignKey in the "many" end and Django automatically creates reverse-mapping.

To do this:
- Create `post` field on Upvote model that is a ForeignKey
- To get all upvotes for a post call `post.upvote_set.all()`. The `upvote_set` is a "magical" variable name that is automatically generated to get all upvotes that reference this post.

**Comment by yiwan** [comment]
OK, doesn't require a lot work as I thought. 🙂

File: lightning_talk/models/post.py (hunk 2 above)
Comment (JuhaS) [inline lightning_talk/models/post.py]: I see you fixed this in new commit, that was fast :). Looking good.

**Comment by yiwan** [comment]
Any advice about the front-end part?

**Comment by makara** [comment]
Thanks @yiwan. It should be enough for us to take a decision. We'll get back to you soon. Closing.

**Pull request closed without merging**